### PR TITLE
nvme-gen-tls-key: add options to derive a TLS key

### PR DESCRIPTION
**Pull request by hreinecke**
Add options to derive a TLS key and store it into the given keyring.

### CODE REVIEW

File: nvme.c [hunk 1]
```diff
@@ -8858,16 +8871,32 @@ static int gen_tls_key(int argc, char **argv, struct command *command, struct pl
...
+		raw_secret[key_len++] = (crc >> 16) & 0xff;
+		raw_secret[key_len++] = (crc >> 24) & 0xff;
 
-	memset(encoded_key, 0, sizeof(encoded_key));
-	base64_encode(raw_secret, key_len, encoded_key);
+		memset(encoded_key, 0, sizeof(encoded_key));
+		base64_encode(raw_secret, key_len, encoded_key);
+		printf("NVMeTLSkey-1:%02x:%s:\n",
+		       cfg.hmac, encoded_key);
+		return 0;
```
Comment (sagi-lb) [inline nvme.c:8885]: I don't understand the semantics of this function. if host and subsys nqns not given you just print the key and exit, and if they were given then it is stored in a keyring?

Comment (hreinecke) [reply nvme.c:8885]: Correct. The 'NVMeTLSkey' thingie is the so-called 'PSK Interchange Format' (cf section 3.6.1.5 of the NVMe TCP spec), and is meant to be used if you want to generate a PSK to be installed on (other) machines.
This format has _no_ indication to which NQN is should match.
From this the actual TLS key should be derived (cf section 3.6.1.3 of the NVMe TCP spec), for which you have to specify the host NQN and subsysten NQN as they are factored in when deriving the TLS key.

File: Documentation/nvme-gen-tls-key.txt [hunk 2]
```diff
@@ -20,6 +23,19 @@ and prints it to stdout.
 
 OPTIONS
 -------
+-k <keyring-name>::
```
Comment (igaw) [inline Documentation/nvme-gen-tls-key.txt:26]: I think the new feature deserves some explanation in the description sections.

File: nvme.c [hunk 3]
```diff
@@ -8858,16 +8871,32 @@ static int gen_tls_key(int argc, char **argv, struct command *command, struct pl
...
+	}
+	if ((!cfg.hostnqn && cfg.subsysnqn) ||
+	    (cfg.hostnqn && !cfg.subsysnqn)) {
```
Comment (igaw) [inline nvme.c:8888]: What about `!(a && b)`?

**Review by igaw — CHANGES_REQUESTED** [review]
I think with a bit documentation this good to go.

**Comment by hreinecke** [comment]
Superseded by PR https://github.com/linux-nvme/nvme-cli/pull/1875 .

**Pull request closed without merging**